astropy/cosmology/tests/test_core.py

```diff
@@ -0,0 +1,325 @@
+
+        assert Example.param is ExampleBase.param
+
+    def test_parameter_doesnt_change_with_cosmology(self, cosmo_cls):
+        """Cosmology reinitializes all descriptors when a subclass is defined."""
+
+        # define subclass to show param is different
+        class Example(cosmo_cls): pass
+
+        assert Example.param is cosmo_cls.param
+
+        # unregister
+        _COSMOLOGY_CLASSES.pop(Example.__qualname__)
+        assert Example.__qualname__ not in _COSMOLOGY_CLASSES
+
+
+# ========================================================================
+
+
+class ParameterTestMixin:
+    """Tests for a Parameter on a Cosmology."""
+
+    def test_Parameters(self, cosmo_cls):
+        """Test `astropy.cosmology.Parameter` attached to Cosmology."""
+        # first establish has expected attribute
+        assert hasattr(cosmo_cls, "__parameters__")
+
+        # check that each entry is a Parameter
+        for n in cosmo_cls.__parameters__:
+            assert hasattr(cosmo_cls, n)  # checks on the instance
+            assert isinstance(getattr(cosmo_cls, n), Parameter)
+
+        # the reverse: check that if it is a Parameter, it's listed.
+        for n in dir(cosmo_cls):
+            if isinstance(getattr(cosmo_cls, n), Parameter):
+                assert n in cosmo_cls.__parameters__
+
+    def test_Parameter_not_unique(self, cosmo_cls, clean_registry):
+        """Cosmology reinitializes Parameter when a class is defined."""
+
+        # define subclass to show param is different
+        class ExampleBase(cosmo_cls):
+            param = Parameter()
+
+        class Example(ExampleBase): pass
+
+        assert Example.param is ExampleBase.param
+        assert Example.__parameters__ == ExampleBase.__parameters__
+
+    def test_Parameters_reorder_by_signature(self, cosmo_cls, clean_registry):
+        """Test parameters are reordered."""
+
+        class Example(cosmo_cls):
+            param = Parameter()
+
+            def __init__(self, param, *, name=None, meta=None):
+                pass  # never actually initialized
+
+        # param should be 1st, all other parameters next
+        Example.__parameters__[0] == "param"
+
+    def test_make_from_Parameter(self, cosmo_cls, clean_registry):
+        """Test the parameter creation process."""
+
+        class Example(cosmo_cls):
+            param = Parameter(unit=u.eV, equivalencies=u.mass_energy())
+
+            def __init__(self, param, *, name=None, meta=None):
+                cls = self.__class__
+                with u.add_enabled_equivalencies(cls.param.equivalencies):
+                    self._param = param << cls.param.unit
+
+        assert Example(1).param == 1 * u.eV
+        assert Example(1 * u.eV).param == 1 * u.eV
+        assert Example(1 * u.J).param == (1 * u.J).to(u.eV)
+        assert Example(1 * u.kg).param == (1 * u.kg).to(u.eV, u.mass_energy())
+
+
+class TestCosmology(ParameterTestMixin, metaclass=abc.ABCMeta):
```

Comment:
Why the metaclass?

Reply:
I'm slowly refactoring of the test suite; and the new ``test_core`` and ``test_flrw`` are part of this effort. The long term goal is to provide inheritable classes for affiliate packages to test a custom Cosmology. I want the structure of the test code to closely mirror the code itself.

e.g. after defining a ``Flatw0wzCDM`` (which doesn't exist)
Writing the test suite will be as simple as 

```python
class TestFlatw0wzCDM(FlatFLRWMixinTest, Testw0waCDM):
    def setup_class(self):
        """Setup for testing."""
        self.cls = Flatw0wzCDM
        self.cls_args = (70 * (u.km / u.s / u.Mpc), 0.27)  # H0, Om0
        self.cls_kwargs = dict(Tcmb0=3 * u.K, name="test", meta={"a": "b"})

    ...  # tests for anything new defined in ``Flatw0wzCDM``
```